src/lib.rs

```diff
@@ -29,15 +30,59 @@ mod dlmalloc;
 #[cfg(all(feature = "global", not(test)))]
 mod global;
 
+/// A platform interface
+pub trait System: Send {
+    /// Allocates a memory region of `size` bytes
+    unsafe fn alloc(size: usize) -> (*mut u8, usize, u32);
+
+    /// Remaps a memory region
+    unsafe fn remap(ptr: *mut u8, oldsize: usize, newsize: usize, can_move: bool) -> *mut u8;
+
+    /// Frees a part of a memory region
+    unsafe fn free_part(ptr: *mut u8, oldsize: usize, newsize: usize) -> bool;
+
+    /// Frees an entire memory region
+    unsafe fn free(ptr: *mut u8, size: usize) -> bool;
+
+    /// Indicates if the platform can release a part of memory
+    fn can_release_part(flags: u32) -> bool;
+
+    /// Allocates a memory region of zeros
+    fn allocates_zeros() -> bool;
+
+    /// Returns the page size
+    fn page_size() -> usize;
+}
+
+/// A platform interface for platforms that support the global lock
+#[cfg(feature = "global")]
+pub trait GlobalSystem: System {
+    /// Acquires the global lock
+    fn acquire_global_lock();
+
+    /// Releases the global lock
+    fn release_global_lock();
+}
+
+/// An allocator instance
+///
+/// Instances of this type are used to allocate blocks of memory. For best
+/// results only use one of these. Currently doesn't implement `Drop` to release
+/// lingering memory back to the OS. That may happen eventually though!
+#[cfg(any(target_os = "linux", target_arch = "wasm32", target_os = "macos"))]
```

Comment:
This is similar to my comment above, but I would prefer that this structure not have a conditional definition.

Reply:
It only conditionally sets the type parameter default. I don't see how you could have a default (in order not to break existing users) but also have platforms that don't have an in-crate implementation without conditional compilation.

Reply:
Platforms which don't have an implementation in this crate would default to just saying "allocation failed"

Reply:
I think this `#[cfg]` can be removed?